File: verdeciel/Location.swift
import UIKit
import QuartzCore
import SceneKit
import Foundation

class Location : Event
{
	var system:Systems!
	var code:String!
	
	var angle:Float!
	var align:Float!
	var distance:CGFloat!
	
	var inCollision:Bool = false
	var inApproach:Bool = false
	var inDiscovery:Bool = false
	var inSight:Bool = false
	
	var mapRequirement:Item!
	
	var isTargetable:Bool = true
	var isTargeted:Bool = false
	var isKnown:Bool = false
	var isSeen:Bool = false
	var isSelected:Bool = false
	var isComplete:Bool! = nil
	
	var structure:Structure!
	var icon:Icon!
	var connection:Location!
	
	var storage:Array<SCNPort> = []
	var isPortEnabled:Bool = false
	
	init(name:String = "",system:Systems = .unknown, at: CGPoint, icon:Icon, structure:Structure)
	{
		super.init(name:name, at:at)
		
		self.name = name
		self.details = "unknown"
		self.at = at
		self.system = system
		self.icon = icon
		self.structure = structure
		self.code = "\(system)-\(name)"
		
		geometry = SCNPlane(width: 0.5, height: 0.5)
		geometry?.firstMaterial?.diffuse.contents = clear
	
		addChildNode(icon)
		
		structure.addHost(self)
		icon.addHost(self)
	}
	
	// MARK: System -
	
	override func whenStart()
	{
		super.whenStart()
		
		position = SCNVector3(at.x,at.y,0)
		distance = distanceBetweenTwoPoints(capsule.at, point2: at)
		angle = calculateAngle()
		align = calculateAlignment()
		
		icon.onUpdate()
	}
	
	func setup()
	{
	
	}
	
	func refresh()
	{
		
	}
	
	override func whenRenderer()
	{
		super.whenRenderer()
		
		position = SCNVector3(at.x,at.y,0)
		distance = distanceBetweenTwoPoints(capsule.at, point2: at)
		angle = calculateAngle()
		align = calculateAlignment()
		
		if distance <= settings.sight { if inSight == false { onSight() ; inSight = true ; isSeen = true } ; sightUpdate() }
		else{ inSight = false }
		
		if distance <= settings.approach { if inApproach == false { inApproach = true ; onApproach() } ; approachUpdate() }
		else{ inApproach = false }
		
		if distance <= settings.collision { if inCollision == false {  inCollision = true ; onCollision() } }
		else{ inCollision = false }
		
		if capsule.isDocked == true && capsule.dock == self { dockUpdate() }
		
		radarCulling()
		clean()
	}
	
	func onSight()
	{
		isSeen = true
		update()
		structure.onSight()
		icon.onUpdate()
	}
	
	func onApproach()
	{
		if mapRequirement != nil && map.port.hasEvent(mapRequirement) == false { return }
		space.startInstance(self)
		// Don't try to dock if there is already a target
		if radar.port.hasEvent() == true && radar.port.event == self || capsule.isFleeing == true {
			capsule.dock(self)
		}
		else if radar.port.hasEvent() == false {
			capsule.dock(self)
		}
		update()
	}
	
	func onCollision()
	{
		update()
	}
	
	func onDock()
	{
		isKnown = true
		update()
		structure.onDock()
		icon.onUpdate()
		exploration.refresh()
		audio.playSound("beep2")
	}
	
	func onUndock()
	{
		retrieveStorage()
		structure.onUndock()
		exploration.refresh()
	}
	
	func retrieveStorage()
	{
		if storage.count == 0 { return }
		
		for port in storage {
			if port.hasItem() == true {
				cargo.addItem(port.event as! Item)
				port.removeEvent()
			}
		}
	}
	
	func onComplete()
	{
		self.isComplete = true
		progress.refresh()
		icon.onUpdate()
		structure.onComplete()
		intercom.complete()
		audio.playSound("beep1")
	}
	
	func sightUpdate()
	{
		structure.sightUpdate()
	}
	
	func approachUpdate()
	{
		
	}
	
	func collisionUpdate()
	{
		
	}
	
	func dockUpdate()
	{
		structure.dockUpdate()
	}
	
	func radarCulling()
	{
		let verticalDistance = abs(capsule.at.y - at.y)
		let horizontalDistance = abs(capsule.at.x - at.x)
		
		// In Sight
		if Float(verticalDistance) <= 1.5 && Float(horizontalDistance) <= 1.5 || radar.overviewMode == true && distance < 7{
			if self.mapRequirement == nil {
				self.show()
			}
			else if self.mapRequirement != nil {
				self.opacity = (map.hasMap(self.mapRequirement) == true && battery.isMapPowered() == true) ? 1 : 0
			}
		}
		// Out Of Sight
		else{
			self.hide()
		}
		
		// Connections
		if connection != nil {
			if connection.opacity == 1 {
				icon.wire.show()
			}
			else{
				icon.wire.hide()
			}
		}
	}
	
	func connect(location:Location)
	{
		connection = location
		icon.wire.update([SCNVector3(0,0,0), SCNVector3( (connection.at.x - self.at.x),(connection.at.y - self.at.y),0)], color: grey)
	}

	// MARK: Events -
	
	override func touch(id:Int)
	{
		if isTargetable == false { return }
		
		if radar.port.event == nil {
			radar.addTarget(self)
			audio.playSound("click3")
		}
		else if radar.port.event == self {
			radar.removeTarget()
			audio.playSound("click2")
		}
		else{
			radar.addTarget(self)
			audio.playSound("click1")
		}
	}
	
	func calculateAngle() -> Float
	{
		let p1 = capsule.at
		let p2 = self.at
		let center = capsule.at
		
		let v1 = CGVector(dx: p1.x - center.x, dy: p1.y - center.y)
		let v2 = CGVector(dx: p2.x - center.x, dy: p2.y - center.y)
		
		let angle = atan2(v2.dy, v2.dx) - atan2(v1.dy, v1.dx)
		
		return (360 - (radToDeg(Float(angle)) - 90)) % 360
	}
	
	func calculateAlignment(direction:Float = capsule.direction) -> Float
	{
		var diff = max(direction, self.angle) - min(direction, self.angle)
		if (diff > 180){ diff = 360 - diff }
		
		return diff
	}
	
	// MARK: Storage -
	
	func storedItems() -> Array<Event>
	{
		var collection:Array<Event> = []
		for port in storage{
			if port.hasEvent() == true && port.event.isQuest == true { collection.append(port.event) }
		}
		return collection
	}
	
	func isStorageEmpty() -> Bool
	{
		for port in storage{
			if port.hasEvent() == true { return true}
		}
		return false
	}
	
	override func payload() -> ConsolePayload
	{
		return ConsolePayload(data:[
			ConsoleData(text: "Name", details: "\(name!)"),
			ConsoleData(text: "System", details: "\(system)"),
			ConsoleData(text: "Position", details: "\(Int(at.x)),\(Int(at.y))"),
			ConsoleData(text: "Distance", details: "\(Int(distance))"),
			ConsoleData(text: "Angle", details: "\(Int(angle))"),
			ConsoleData(text: details)
			])
	}
	
	func close()
	{
		icon.close()
	}
	
	required init(coder aDecoder: NSCoder)
	{
		fatalError("init(coder:) has not been implemented")
	}
}

File: verreciel_js/scripts/locations/location.js
class Location extends Event
{
  constructor(name, system, at, icon, structure)
  {
    // assertArgs(arguments, 5);
    super(name, at, "unknown", verreciel.grey, false);
    
    this.system = system;
    this.icon = icon;
    this.structure = structure;
    this.code = system + "-" + name;
    this.storage = [];

    this.isDocked = false;
    this.inCollision = false;
    this.inApproach = false;
    this.inDiscovery = false;
    this.inSight = false;
    this.isTargetable = true;
    this.isTargeted = false;
    this.isKnown = false;
    this.isSeen = false;
    this.isSelected = false;
    this.isComplete = null;
    this.isPortEnabled = false;
    
    this.add(this.icon);
    
    this.structure.addHost(this);
    this.icon.addHost(this);
  }
  
  // MARK: System -
  
  whenStart()
  {
    // assertArgs(arguments, 0);
    super.whenStart();
    
    this.position.set(this.at.x,this.at.y,0);
    this.distance = distanceBetweenTwoPoints(verreciel.capsule.at, this.at);
    this.angle = this.calculateAngle();
    this.align = this.calculateAlignment();
    this.icon.onUpdate();
  }
  
  setup()
  {
    // assertArgs(arguments, 0);
  
  }
  
  refresh()
  {
    // assertArgs(arguments, 0);
    
  }
  
  whenRenderer()
  {
    // assertArgs(arguments, 0);
    super.whenRenderer();
    
    this.position.set(this.at.x,this.at.y,0);
    this.distance = distanceBetweenTwoPoints(verreciel.capsule.at, this.at);
    this.angle = this.calculateAngle();
    this.align = this.calculateAlignment();
    
    if (this.distance <= Settings.sight)
    {
      if (this.inSight == false)
      { 
        this.onSight(); 
        this.inSight = true;
        this.isSeen = true;
      }
      this.sightUpdate();
    }
    else
    {
      this.inSight = false;
    }
    
    if (this.distance <= Settings.approach)
    {
      if (this.inApproach == false)
      {
        this.inApproach = true;
        this.onApproach();
      }
      this.approachUpdate();
    }
    else
    {
      this.inApproach = false;
    }
    
    if (this.distance <= Settings.collision)
    {
      if (this.inCollision == false)
      {
        this.inCollision = true;
        this.onCollision();
      }
    }
    else
    {
      this.inCollision = false;
    }
    
    if (verreciel.capsule.isDocked == true && verreciel.capsule.location == this)
    {
      this.dockUpdate();
    }
    
    this.radarCulling();
    this.clean();
  }
  
  onRadarView()
  {
    // assertArgs(arguments, 0);
    this.icon.label.opacity = 1;
  }
  
  onHelmetView()
  {
    // assertArgs(arguments, 0);
    this.icon.label.opacity = 0;
  }
  
  onSight()
  {
    // assertArgs(arguments, 0);
    this.isSeen = true;
    this.update();
    this.icon.onUpdate();

    if (this.isDocked)
    {
      this.structure.onDock();
    }
    else
    {
      this.structure.onSight();
    }
  }
  
  onApproach()
  {
    // assertArgs(arguments, 0);
    if (this.mapRequirement != null && verreciel.nav.port.hasEvent(this.mapRequirement) == false)
    {
      return;
    }
    verreciel.space.startInstance(this);
    // Don't try to dock if there is already a target
    if (verreciel.radar.port.hasEvent() == true && verreciel.radar.port.event == this || verreciel.capsule.isFleeing == true)
    {
      verreciel.capsule.dock(this);
    }
    else if (verreciel.radar.port.hasEvent() == false)
    {
      verreciel.capsule.dock(this);
    }
    this.update();
  }
  
  onCollision()
  {
    // assertArgs(arguments, 0);
    this.update();
  }
  
  onDock()
  {
    // assertArgs(arguments, 0);
    if (verreciel.thruster.isLocked && verreciel.universe.loiqe_city.isKnown == true)
    {
      verreciel.thruster.unlock();
    }
    
    this.isDocked = true;
    this.isKnown = true;
    this.update();
    this.structure.onDock();
    this.icon.onUpdate();
    verreciel.exploration.refresh();
    verreciel.music.playEffect("beep2");
  }
  
  onUndock()
  {
    // assertArgs(arguments, 0);
    this.isDocked = false;
    this.retrieveStorage();
    this.structure.onUndock();
    verreciel.exploration.refresh();
  }
  
  retrieveStorage()
  {
    // assertArgs(arguments, 0);
    if (this.storage.length == 0)
    {
      return;
    }
    
    for (let port of this.storage)
    {
      if (port.hasItem() == true)
      {
        verreciel.cargo.addItem(port.event);
        port.removeEvent();
      }
    }
  }
  
  onComplete()
  {
    // assertArgs(arguments, 0);
    this.isComplete = true;
    verreciel.progress.refresh();
    this.icon.onUpdate();
    this.structure.onComplete();
    verreciel.intercom.complete();
    verreciel.music.playEffect("beep1");
  }
  
  sightUpdate()
  {
    // assertArgs(arguments, 0);
    this.structure.sightUpdate();
  }
  
  approachUpdate()
  {
    // assertArgs(arguments, 0);
    
  }
  
  collisionUpdate()
  {
    // assertArgs(arguments, 0);
    
  }
  
  dockUpdate()
  {
    // assertArgs(arguments, 0);
    this.structure.dockUpdate();
  }
  
  radarCulling()
  {
    // assertArgs(arguments, 0);
    let verticalDistance = Math.abs(verreciel.capsule.at.y - this.at.y);
    let horizontalDistance = Math.abs(verreciel.capsule.at.x - this.at.x);
    
    // In Sight
    if (verticalDistance <= 1.5 && horizontalDistance <= 1.5 || verreciel.radar.overviewMode == true && this.distance < 7)
    {
      if (this.mapRequirement == null)
      {
        this.show();
      }
      else if (this.mapRequirement != null)
      {
        this.opacity = (verreciel.nav.hasMap(this.mapRequirement) == true && verreciel.battery.isNavPowered() == true) ? 1 : 0;
      }
    }
    // Out Of Sight
    else
    {
      this.hide();
    }
    
    // Connections
    if (this.connection != null)
    {
      if (this.connection.opacity != 0)
      {
        this.icon.wire.show();
      }
      else
      {
        this.icon.wire.hide();
      }
    }
  }
  
  connect(location)
  {
    // assertArgs(arguments, 1);
    this.connection = location;
    this.icon.wire.updateVertices([
      new THREE.Vector3(0,0,0), 
      new THREE.Vector3( (this.connection.at.x - this.at.x),(this.connection.at.y - this.at.y),0)
    ]);
  }

  // MARK: Events -
  
  touch(id)
  {
    // assertArgs(arguments, 1);
    if (this.isTargetable == false)
    {
      return false;
    }
    if (verreciel.thruster.isLocked == true && verreciel.game.state > 2)
    {
      return false;
    }
    if (verreciel.radar.port.event == null)
    {
      verreciel.radar.addTarget(this);
      verreciel.music.playEffect("click3");
    }
    else if (verreciel.radar.port.event == this)
    {
      verreciel.radar.removeTarget();
      verreciel.music.playEffect("click2");
    }
    else
    {
      verreciel.radar.addTarget(this);
      verreciel.music.playEffect("click1");
    }
    if (verreciel.radar.port.isConnectedToPanel(verreciel.console) == true)
    {
      verreciel.console.onConnect();
    }
    return true;
  }
  
  calculateAngle()
  {
    // assertArgs(arguments, 0);
    let angle = angleBetweenTwoPoints(verreciel.capsule.at, this.at, verreciel.capsule.at);
    return (360 - (angle - 90)) % 360;
  }
  
  calculateAlignment(direction = verreciel.capsule.direction)
  {
    // assertArgs(arguments, 0);
    var diff = Math.max(direction, this.angle) - Math.min(direction, this.angle);
    if (diff > 180)
    {
      diff = 360 - diff;
    }
    
    return diff;
  }
  
  // MARK: Storage -
  
  storedItems()
  {
    // assertArgs(arguments, 0);
    var collection = [];
    for (let port of this.storage)
    {
      if (port.hasEvent() == true && port.event.isQuest == true)
      {
        collection.push(port.event);
      }
    }
    return collection;
  }
  
  payload()
  {
    // assertArgs(arguments, 0);
    return new ConsolePayload([
      new ConsoleData("Name", this.name),
      new ConsoleData("System", this.system),
      new ConsoleData("Position", this.at.x.toFixed(0) + "," + this.at.y.toFixed(0)),
      new ConsoleData("Distance", this.distance.toFixed(0)),
      new ConsoleData("Angle", this.angle.toFixed(0)),
      new ConsoleData(this.details),
    ]);
  }
  
  close()
  {
    // assertArgs(arguments, 0);
    this.icon.close();
  }
}
